subsys/bluetooth/host/l2cap.c

```diff
@@ -2267,6 +2292,95 @@ static void l2cap_chan_send_credits(struct bt_l2cap_le_chan *chan,
 	LOG_DBG("chan %p credits %lu", chan, atomic_get(&chan->rx.credits));
 }
 
+#ifdef CONFIG_BT_L2CAP_SEG_RECV
+static int l2cap_chan_send_credits_pdu(struct bt_conn *conn, uint16_t cid, uint16_t credits)
+{
+	int err;
+	struct net_buf *buf;
+	struct bt_l2cap_le_credits *ev;
+
+	buf = l2cap_create_le_sig_pdu(NULL, BT_L2CAP_LE_CREDITS, get_ident(), sizeof(*ev));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	ev = net_buf_add(buf, sizeof(*ev));
+	*ev = (struct bt_l2cap_le_credits){
```

Comment:
Do you really need the explicit typecast? I'd have thought that the type gets inferred from the type of `*ev`.

Reply:
That said, I'm not sure there's much advantage of doing the initialisation like this, especially with just two struct members. I believe most of the subsystem uses a style like this:
```c
ev->cid = sys_cpu_to_le16(cid);
ev->credits = sys_cpu(to_le16(credits);
```

Reply:
I assume that doing it this way will ensure that any other fields are set to `0`, similar to "regular" struct initialization?

Reply:
It's simpler than it looks. I'm happy to explain :)

> Do you really need the explicit typecast? I'd have thought that the type gets inferred from the type of `*ev`.

This is not a declaration, so the special declaration syntax (`= {...}`) does not apply. This is simply a combined use of an object literal, struct assignment and pointer de-referencing. Object literals always have the form `(type){initializer-list}`.

There is no 'downwards' type inference in C. The object literal is a child-node of the assignment. The child-node does not get type information from the parent, so the object literal must know it's own type.

> I assume that doing it this way will ensure that any other fields are set to `0`, similar to "regular" struct initialization?

This is correct. As explained above, this is a normal object literal, and struct assignment is equivalent to assigning all fields of the struct type one-by-one. Struct assignment semantics are not special for object literals; it works just as well for two struct variables `obj1 = obj2;`.

I would love if there was a linter pragma "no fields omitted in initializer", but there was nothing that easy. So omitted fields are just zeroed as you say.

I think there is a slight benefit to zeroing over just not initializing, as it's deterministic and zeros are easier to spot when debugging. Seeing as there is no new language feature here, I hope we can keep this as is, but I don't have any issue changing this to scalar assignments if we prefer that. I value you input on the readability aspect of this :).

Reply:
@alwa-nordic thanks for the explanation. Seems that even after 25 years of coding in C  I’m still not 100% proficient :)

I’m fine keeping it as it is now.